packages/punchcard/lib/appsync/api.ts

```diff
@@ -0,0 +1,324 @@
+import type * as appsync from '@aws-cdk/aws-appsync';
+
+import { Meta, Shape, ShapeGuards } from '@punchcard/shape';
+import { RecordShape } from '@punchcard/shape/lib/record';
+import { Build } from '../core/build';
+import { CDK } from '../core/cdk';
+import { Construct, Scope } from '../core/construct';
+import { Resource } from '../core/resource';
+import { VExpression } from './expression';
+import { ApiFragment } from './fragment';
+import { Statement, StatementGuards } from './statement';
+import { TypeSpec, TypeSystem } from './type-system';
+import { VTL } from './vtl';
+import { VObject } from './vtl-object';
+
+export interface OverrideApiProps extends Omit<appsync.GraphQLApiProps,
+  | 'name'
+  | 'schemaDefinition'
+  | 'schemaDefinitionFile'
+> {}
+
+export interface ApiProps<
+  T extends TypeSystem,
+  Q extends keyof T | undefined = undefined,
+  M extends keyof T | undefined = undefined
+> {
+  types: ApiFragment<T>;
+  query?: Q;
+  mutation?: M;
+  name: string;
+  overrideProps?: Build<OverrideApiProps>;
+}
+
+/**
+ * A finalized AppSync-managed GraphQL API.
+ *
+ * APIs are constructed by combining `ApiFragments`.
+ *
+ * @typeparam Types - map of names to types in this API
+ */
+export class Api<
+  T extends TypeSystem,
+  Q extends keyof T | undefined,
+  M extends keyof T | undefined
+> extends Construct implements Resource<appsync.CfnGraphQLApi> {
+  public readonly resource: Build<appsync.CfnGraphQLApi>;
+
+  public readonly Types: {
+    // eumerate through to clean up the type siganture ("Compaction").
+    [t in keyof T]: {
+      [k in keyof T[t]]: T[t][k]
+    };
+  };
+  public readonly QueryFQN: Q;
+  public readonly QueryType: Q extends keyof T ? T[Q]['type'] : undefined;
+  public readonly MutationFQN: M;
+  public readonly MutationType: M extends keyof T ? T[M]['type'] : undefined;
+
+  constructor(scope: Scope, id: string, props: ApiProps<T, Q, M>) {
+    super(scope, id);
+    this.Types = props.types.Types;
+    if (props.query !== undefined) {
+      this.QueryFQN = props.query;
+      this.QueryType = this.Types[props.query!].type as any;
+    }
+    if (props.mutation !== undefined) {
+      this.MutationFQN = props.mutation;
+      this.MutationType = this.Types[props.mutation!].type as any;
+    }
+
+    this.resource = Build.concat(
```

Comment:
This is the magic - it interprets the Velocity template "programs" and synthesizes a GraphQL Schema and Resolver Pipelines.